# Description

TODO

# Environment variables

In [1]:
# from IPython.display import display

# import conf

# N_JOBS = conf.GENERAL["N_JOBS"]
# display(N_JOBS)

In [2]:
# %env MKL_NUM_THREADS=$N_JOBS
# %env OPEN_BLAS_NUM_THREADS=$N_JOBS
# %env NUMEXPR_NUM_THREADS=$N_JOBS
# %env OMP_NUM_THREADS=$N_JOBS

# Modules loading

In [3]:
%load_ext autoreload
%autoreload 2

In [4]:
from pathlib import Path

import numpy as np
import pandas as pd

import conf
from data.cache import read_data

# Settings

In [5]:
# SELECTED_PARTITION_K = 44

In [6]:
# CONSENSUS_CLUSTERING_DIR = Path(
#     conf.RESULTS["CLUSTERING_DIR"], "consensus_clustering"
# ).resolve()

# display(CONSENSUS_CLUSTERING_DIR)

# Data loading

## eMERGE S-MultiXcan results

In [7]:
# FIXME: path hardcoded
input_filepath = Path(
    "/home/miltondp/projects/labs/greenelab/phenoplier/base/data/emerge/gene_assoc/emerge-smultixcan-mashr-zscores.pkl"
).resolve()

In [8]:
# input_filepath = Path(
#     conf.RESULTS["PROJECTIONS_DIR"],
#     "projection-emerge-smultixcan-mashr-zscores.pkl",
# ).resolve()
# display(input_filepath)
# assert input_filepath.exists()

In [9]:
pmbb_data = pd.read_pickle(input_filepath)

In [10]:
pmbb_data.shape

(22198, 309)

In [11]:
pmbb_data.head()

292.3       079       741     418.1       280       747  \
gene_name                                                                     
ENSG00000000419  1.043436  1.134028  1.595979  0.399451  0.725100  1.655468   
ENSG00000000457  1.236752  0.119837  0.224120  0.200610  1.304190  0.541478   
ENSG00000000460  0.497108  0.005690  0.317710  0.184918  1.136574  1.708245   
ENSG00000000938  0.374442  1.099899  1.975795  0.107667  0.903520  0.673819   
ENSG00000000971  1.108762  1.133163  0.227188  0.222948  0.072709  0.256124   

                      591    426.21       244     070.3  ...       427  \
gene_name                                                ...             
ENSG00000000419  1.803598  0.125250  1.053218  0.922592  ...  0.364789   
ENSG00000000457  0.646474  0.493759  0.189680  1.563273  ...  1.519500   
ENSG00000000460  0.288962  1.070026  0.403510  1.106321  ...  0.678253   
ENSG00000000938  0.855749  0.499088  0.375406  0.815757  ...  1.585687   
ENSG00000000971  0.540911  0.606947  0.229133  1.557433  ...  2.349373   

                    585.4       345     366.2     420.2       195     250.2  \
gene_name                                                                     
ENSG00000000419  0.453873  0.759558  1.173671  0.298944  1.029646  0.826238   
ENSG00000000457  0.201290  1.020996  0.592006  0.632661  0.995453  0.069362   
ENSG00000000460  1.166204  1.138126  0.608565  0.449551  0.096689  0.887467   
ENSG00000000938  0.584143  0.751384  2.758230  0.572712  0.141023  1.437022   
ENSG00000000971  1.122458  0.155346  0.990123  0.326914  1.100210  1.099537   

                    250.1       371       562  
gene_name                                      
ENSG00000000419  0.685711  2.069701  1.544766  
ENSG00000000457  0.460979  0.217046  1.350475  
ENSG00000000460  0.143815  0.236647  1.048859  
ENSG00000000938  0.288943  0.665727  1.225229  
ENSG00000000971  2.511358  0.950262  1.467305  

[5 rows x 309 columns]

## PhenomeXcan S-MultiXcan results

In [12]:
input_filepath = conf.PHENOMEXCAN["SMULTIXCAN_EFO_PARTIAL_MASHR_ZSCORES_FILE"]
display(input_filepath)

PosixPath('/home/miltondp/projects/labs/greenelab/phenoplier/bases_data/base_orig/data/phenomexcan/gene_assoc/smultixcan-efo_partial-mashr-zscores.pkl')

In [13]:
phenomexcan_data = pd.read_pickle(input_filepath)

In [14]:
phenomexcan_data.shape

(22515, 3752)

In [15]:
phenomexcan_data.head()

100001_raw-Food_weight  100002_raw-Energy  \
gene_name                                                    
ENSG00000000419                1.145442           0.724557   
ENSG00000000457                0.618066           1.028131   
ENSG00000000460                0.515724           0.403596   
ENSG00000000938                0.280781           0.253910   
ENSG00000000971                0.548127           0.389877   

                 100003_raw-Protein  100004_raw-Fat  100005_raw-Carbohydrate  \
gene_name                                                                      
ENSG00000000419            0.090876        0.298165                 1.134347   
ENSG00000000457            2.218420        0.762584                 0.934418   
ENSG00000000460            1.251359        0.433091                 0.413466   
ENSG00000000938            0.879148        0.352705                 0.051846   
ENSG00000000971            0.723469        1.167250                 0.315952   

                 100006_raw-Saturated_fat  100007_raw-Polyunsaturated_fat  \
gene_name                                                                   
ENSG00000000419                  1.371138                        0.065718   
ENSG00000000457                  0.192993                        1.080230   
ENSG00000000460                  0.246261                        1.236151   
ENSG00000000938                  0.184212                        0.148566   
ENSG00000000971                  0.324939                        1.613932   

                 100008_raw-Total_sugars  100009_raw-Englyst_dietary_fibre  \
gene_name                                                                    
ENSG00000000419                 0.794317                          0.600342   
ENSG00000000457                 0.765997                          0.375898   
ENSG00000000460                 0.827430                          0.571985   
ENSG00000000938                 0.009989                          0.363751   
ENSG00000000971                 0.311432                          0.333548   

                 100010-Portion_size  ...  visual impairment  vitiligo  \
gene_name                             ...                                
ENSG00000000419             0.317652  ...           0.360518  1.351624   
ENSG00000000457             0.678731  ...           2.134504  0.127830   
ENSG00000000460             0.782174  ...           1.768905  0.992408   
ENSG00000000938             0.374514  ...           0.656552  2.046041   
ENSG00000000971             1.807243  ...           0.260482  0.646204   

                 vitreous body disease  vocal cord polyp  voice disorders  \
gene_name                                                                   
ENSG00000000419               1.157695          0.835289         1.173072   
ENSG00000000457               0.534690          0.120516         0.517464   
ENSG00000000460               0.548215          0.412341         1.499415   
ENSG00000000938               2.746832          0.108211         1.008258   
ENSG00000000971               1.080240          0.678330         1.465358   

                 wellbeing measurement AND family relationship  wheezing  \
gene_name                                                                  
ENSG00000000419                                       1.337280  1.743822   
ENSG00000000457                                       2.545363  0.673331   
ENSG00000000460                                       1.366780  0.443318   
ENSG00000000938                                       0.755695  0.896228   
ENSG00000000971                                       0.307672  0.118376   

                 whooping cough  worry measurement  wrist fracture  
gene_name                                                           
ENSG00000000419        1.017226           1.512184        0.972241  
ENSG00000000457        2.003092           0.344000        2.033122  
ENSG00000000460        0.417630           0.225934        1.613246

## PhenomeXcan traits info

In [16]:
phenomexcan_traits_info = read_data(conf.PHENOMEXCAN["RAPID_GWAS_PHENO_INFO_FILE"])

In [17]:
phenomexcan_traits_info.shape

(4359, 9)

In [18]:
phenomexcan_traits_info.head()

description variable_type  \
phenotype                                                                    
G54        Diagnoses - main ICD10: G54 Nerve root and ple...   categorical   
C21        Diagnoses - main ICD10: C21 Malignant neoplasm...   categorical   
F45         Diagnoses - main ICD10: F45 Somatoform disorders   categorical   
I35        Diagnoses - main ICD10: I35 Nonrheumatic aorti...   categorical   
D38        Diagnoses - main ICD10: D38 Neoplasm of uncert...   categorical   

          source  n_non_missing  n_missing  n_controls  n_cases  \
phenotype                                                         
G54        icd10         361194          0    361051.0    143.0   
C21        icd10         361194          0    361055.0    139.0   
F45        icd10         361194          0    360953.0    241.0   
I35        icd10         361194          0    360293.0    901.0   
D38        icd10         361194          0    361087.0    107.0   

          PHESANT_transformation notes  
phenotype                               
G54                          NaN   NaN  
C21                          NaN   NaN  
F45                          NaN   NaN  
I35                          NaN   NaN  
D38                          NaN   NaN

In [19]:
# read_data(conf.PHENOMEXCAN["GTEX_GWAS_PHENO_INFO_FILE"])

## PMBB/eMERGE phecodes mappings to ICD10

In [20]:
# FIXME: hardcoded
# input_file = conf.PMBB["ICD10_PHECODE_MAPPING_FILE"]
input_file = Path(
    "/home/miltondp/projects/labs/greenelab/phenoplier/base/data/pmbb",
    "Phecode_to_UKB_ICD10_mapping.tsv",
).resolve()
display(input_file)

PosixPath('/home/miltondp/projects/labs/greenelab/phenoplier/bases_data/base_orig/data/pmbb/Phecode_to_UKB_ICD10_mapping.tsv')

In [21]:
icd10_to_phecodes_map = pd.read_csv(input_file, sep="\t", dtype={"Phecode": str})

In [22]:
icd10_to_phecodes_map.shape

(235, 4)

In [23]:
icd10_to_phecodes_map.head()

PhenomeXcan_ICD10                      PhenomeXcan_ICD10_Description  \
0               A09  Diarrhea and gastroenteritis of infectious origin   
1               A01                     Typhoid and paratyphoid fevers   
2               A02                        Other salmonella infections   
3               A03                                        Shigellosis   
4               A04              Other bacterial intestinal infections   

  Phecode   Phecode_Description  
0     008  Intestinal infection  
1   008.5   Bacterial enteritis  
2   008.5   Bacterial enteritis  
3   008.5   Bacterial enteritis  
4   008.5   Bacterial enteritis

In [24]:
icd10_to_phecodes_map["Phecode"].is_unique

False

In [25]:
icd10_to_phecodes_map["PhenomeXcan_ICD10"].is_unique

False

The same ICD10 code from PhenomeXcan can map to multiple phecodes, and one phecode can belong to several ICD10 codes.

## Phecodes descriptions (and sample size)

In [26]:
# FIXME: hardcoded
input_filepath = Path(
    "/home/miltondp/projects/labs/greenelab/phenoplier/base/data/emerge",
    "eMERGE_III_PMBB_GSA_v2_2020_phecode_AFR_EUR_cc50_counts_w_dictionary.txt",
).resolve()
display(input_filepath)

PosixPath('/home/miltondp/projects/labs/greenelab/phenoplier/bases_data/base_orig/data/emerge/eMERGE_III_PMBB_GSA_v2_2020_phecode_AFR_EUR_cc50_counts_w_dictionary.txt')

In [27]:
pmbb_traits_df = pd.read_csv(
    input_filepath,
    sep="\t",
    dtype={"phecode": str},
    usecols=["phecode", "phenotype", "category"],
)

In [28]:
pmbb_traits_df = pmbb_traits_df.rename(
    columns={
        "phenotype": "phecode_phenotype",
        "category": "phecode_category",
    }
)

In [29]:
pmbb_traits_df.shape

(309, 3)

In [30]:
pmbb_traits_df.head()

phecode                         phecode_phenotype     phecode_category
0     008                      Intestinal infection  infectious diseases
1   008.5                       Bacterial enteritis  infectious diseases
2  008.52  Intestinal infection due to C. difficile  infectious diseases
3     038                                Septicemia  infectious diseases
4   038.3                                Bacteremia  infectious diseases

## PhenomeXcan-EFO mappings

In [31]:
phenomexcan_to_efo = read_data(conf.PHENOMEXCAN["TRAITS_FULLCODE_TO_EFO_MAP_FILE"])

In [32]:
phenomexcan_to_efo.shape

(1087, 7)

In [33]:
phenomexcan_to_efo.head()

ukb_code              term_label                term_codes mapping_type  \
0      K55        vascular disease  EFO:0004264, EFO:0009431        Broad   
1      M17  osteoarthritis || knee               EFO:0004616        Broad   
2      R30                 dysuria               EFO:0003901      ? Broad   
3      O60         premature birth               EFO:0003917      ? Exact   
4      S64  carpal tunnel syndrome               EFO:0004143     ? Narrow   

                                        ukb_fullcode  \
0  K55-Diagnoses_main_ICD10_K55_Vascular_disorder...   
1  M17-Diagnoses_main_ICD10_M17_Gonarthrosis_arth...   
2  R30-Diagnoses_main_ICD10_R30_Pain_associated_w...   
3      O60-Diagnoses_main_ICD10_O60_Preterm_delivery   
4  S64-Diagnoses_main_ICD10_S64_Injury_of_nerves_...   

                        current_term_label category  
0  vascular disease AND intestinal disease  disease  
1                     osteoarthritis, knee  disease  
2                                  dysuria      NaN  
3                          premature birth      NaN  
4                   carpal tunnel syndrome  disease

# How many PhenomeXcan traits are in eMERGE results?

In [34]:
phenomexcan_icd10_traits = phenomexcan_traits_info[
    phenomexcan_traits_info["source"] == "icd10"
]

In [35]:
phenomexcan_icd10_traits.shape

(633, 9)

In [36]:
phenomexcan_icd10_traits.head()

description variable_type  \
phenotype                                                                    
G54        Diagnoses - main ICD10: G54 Nerve root and ple...   categorical   
C21        Diagnoses - main ICD10: C21 Malignant neoplasm...   categorical   
F45         Diagnoses - main ICD10: F45 Somatoform disorders   categorical   
I35        Diagnoses - main ICD10: I35 Nonrheumatic aorti...   categorical   
D38        Diagnoses - main ICD10: D38 Neoplasm of uncert...   categorical   

          source  n_non_missing  n_missing  n_controls  n_cases  \
phenotype                                                         
G54        icd10         361194          0    361051.0    143.0   
C21        icd10         361194          0    361055.0    139.0   
F45        icd10         361194          0    360953.0    241.0   
I35        icd10         361194          0    360293.0    901.0   
D38        icd10         361194          0    361087.0    107.0   

          PHESANT_transformation notes  
phenotype                               
G54                          NaN   NaN  
C21                          NaN   NaN  
F45                          NaN   NaN  
I35                          NaN   NaN  
D38                          NaN   NaN

In [37]:
# See how many traits
icd10_in_phecodes = icd10_to_phecodes_map[
    icd10_to_phecodes_map["PhenomeXcan_ICD10"].isin(phenomexcan_icd10_traits.index)
]

In [38]:
icd10_in_phecodes.shape

(224, 4)

224 PhenomeXcan's traits from ICD10 have phecodes

In [39]:
icd10_in_phecodes.head()

PhenomeXcan_ICD10                      PhenomeXcan_ICD10_Description  \
0                A09  Diarrhea and gastroenteritis of infectious origin   
4                A04              Other bacterial intestinal infections   
6                A41                                   Other septicemia   
9                B02                             Zoster [herpes zoster]   
10               B18                            Chronic viral hepatitis   

   Phecode   Phecode_Description  
0      008  Intestinal infection  
4    008.5   Bacterial enteritis  
6      038            Septicemia  
9      053         Herpes zoster  
10     070       Viral hepatitis

In [40]:
# Which entries in the icd10-phecode mapping are not present in PhenomeXcan?
icd10_to_phecodes_map[
    ~icd10_to_phecodes_map["PhenomeXcan_ICD10"].isin(phenomexcan_icd10_traits.index)
]

PhenomeXcan_ICD10                      PhenomeXcan_ICD10_Description  \
1                 A01                     Typhoid and paratyphoid fevers   
2                 A02                        Other salmonella infections   
3                 A03                                        Shigellosis   
5                 A05  Other bacterial foodborne intoxications, not e...   
7                 A42                                      Actinomycosis   
8                 B00            Herpesviral [herpes simplex] infections   
29                D20  Benign neoplasm of soft tissue of retroperiton...   
36                C96  Other and unspecified malignant neoplasms of l...   
46                E12           Malnutrition-related diabetes mellitus     
142               J05  Acute obstructive laryngitis [croup] and epigl...   
205               L89                                    Decubitus ulcer   

    Phecode                           Phecode_Description  
1     008.5                           Bacterial enteritis  
2     008.5                           Bacterial enteritis  
3     008.5                           Bacterial enteritis  
5     008.5                           Bacterial enteritis  
7       038                                    Septicemia  
8       038                                    Septicemia  
29      199                Neoplasm of uncertain behavior  
36      202  Cancer of other lymphoid, histiocytic tissue  
46      249                   Secondary diabetes mellitus  
142     483            Acute bronchitis and bronchiolitis  
205   707.1                               Decubitus ulcer

# Map ICD10 to EFO in PhenomeXcan

In [41]:
# get the EFO map of PhenomeXcan's traits with ICD10 that have phecodes
efo_from_icd10 = phenomexcan_to_efo[
    phenomexcan_to_efo["ukb_code"].isin(icd10_in_phecodes["PhenomeXcan_ICD10"].values)
]

In [42]:
efo_from_icd10

ukb_code                                         term_label  \
1         M17                             osteoarthritis || knee   
2         R30                                            dysuria   
7         M05                               rheumatoid arthritis   
9         K74                                 cirrhosis of liver   
10        M15                                     osteoarthritis   
...       ...                                                ...   
1043      C78  respiratory system neoplasm||digestive system ...   
1044      C79                                           neoplasm   
1046      C85                              non-Hodgkins lymphoma   
1047      B37                              Candidiasis, Invasive   
1048      A04              intestinal disease||bacterial disease   

                    term_codes mapping_type  \
1                  EFO:0004616        Broad   
2                  EFO:0003901      ? Broad   
7                  EFO:0000685        Broad   
9                  EFO:0001422        Broad   
10                 EFO:0002506        Broad   
...                        ...          ...   
1043               EFO:0009812        Broad   
1044               EFO:0009812        Broad   
1046               EFO:0005952        Exact   
1047               EFO:1001283       Narrow   
1048  EFO:0009431||EFO:0000771        Broad   

                                           ukb_fullcode  \
1     M17-Diagnoses_main_ICD10_M17_Gonarthrosis_arth...   
2     R30-Diagnoses_main_ICD10_R30_Pain_associated_w...   
7     M05-Diagnoses_main_ICD10_M05_Seropositive_rheu...   
9     K74-Diagnoses_main_ICD10_K74_Fibrosis_and_cirr...   
10           M15-Diagnoses_main_ICD10_M15_Polyarthrosis   
...                                                 ...   
1043  C78-Diagnoses_main_ICD10_C78_Secondary_maligna...   
1044  C79-Diagnoses_main_ICD10_C79_Secondary_maligna...   
1046  C85-Diagnoses_main_ICD10_C85_Other_and_unspeci...   
1047           B37-Diagnoses_main_ICD10_B37_Candidiasis   
1048  A04-Diagnoses_main_ICD10_A04_Other_bacterial_i...   

                            current_term_label category  
1                         osteoarthritis, knee  disease  
2                                      dysuria      NaN  
7                         rheumatoid arthritis  disease  
9                           cirrhosis of liver  disease  
10                              osteoarthritis  disease  
...                                        ...      ...  
1043              secondary malignant neoplasm  disease  
1044              secondary malignant neoplasm  disease  
1046                     non-hodgkins lymphoma  disease  
1047                     candidiasis, invasive  disease  
1048  intestinal disease AND bacterial disease  disease  

[199 rows x 7 columns]

199 PhenomeXcan traits can be mapped to phecodes when we use the PhenomeXcan-EFO mappings. These mappings combine different PhenomeXcan traits, such as ICD10 codes and self-reported data.

# Create initial dataframe for test set

We create a test set with all 309 PMBB's traits available. It is important to include them all, since some ICD10 are mapped to several different phecodes, and one phecodes can represent different ICD10 codes.

In [43]:
test_df = pmbb_traits_df.copy()

In [44]:
test_df.shape

(309, 3)

In [45]:
test_df.head()

phecode                         phecode_phenotype     phecode_category
0     008                      Intestinal infection  infectious diseases
1   008.5                       Bacterial enteritis  infectious diseases
2  008.52  Intestinal infection due to C. difficile  infectious diseases
3     038                                Septicemia  infectious diseases
4   038.3                                Bacteremia  infectious diseases

## Phecode to ICD10

In [46]:
def map_phecode_to_icd10_list(phecode):
    df = icd10_to_phecodes_map[icd10_to_phecodes_map["Phecode"] == phecode]
    if df.shape[0] == 0:
        return None

    return set(df["PhenomeXcan_ICD10"].values)

In [47]:
# some testing
assert map_phecode_to_icd10_list("008") == set(["A09"])
assert map_phecode_to_icd10_list("244.2") == set(["E03"])
assert map_phecode_to_icd10_list("244.4") == set(["E03"])

assert map_phecode_to_icd10_list("216") == set(["D22", "D23"])

assert map_phecode_to_icd10_list("722") is None

In [48]:
_new_df = []

for phecode in test_df["phecode"].unique():
    _new_map_values = map_phecode_to_icd10_list(phecode)
    if _new_map_values is None:
        continue

    for n in _new_map_values:
        _new_df.append({"phecode": phecode, "icd10": n})

_new_df = pd.DataFrame(_new_df)

In [49]:
# assign phecode
test_df = pd.merge(test_df, _new_df, on="phecode", how="left")

In [50]:
test_df.shape

(368, 4)

In [51]:
test_df.head()

phecode     phecode_phenotype     phecode_category icd10
0     008  Intestinal infection  infectious diseases   A09
1   008.5   Bacterial enteritis  infectious diseases   A02
2   008.5   Bacterial enteritis  infectious diseases   A05
3   008.5   Bacterial enteritis  infectious diseases   A03
4   008.5   Bacterial enteritis  infectious diseases   A01

## ICD10 to EFO

In [52]:
def map_icd10_to_efo(icd10_list):
    if icd10_list is None:
        return None

    df = efo_from_icd10[efo_from_icd10["ukb_code"] == icd10_list]
    efo_values = None

    if df.shape[0] > 0:
        efo_values = set(df["current_term_label"].values)

    return efo_values

In [53]:
# some testing
assert map_icd10_to_efo("M17") == set(["osteoarthritis, knee"])
assert map_icd10_to_efo("R30") == set(["dysuria"])
assert map_icd10_to_efo("I49") == set(["cardiac arrhythmia"])
assert map_icd10_to_efo("A41") == set(["sepsis"])

assert map_icd10_to_efo("T81") == set(["complication"])
assert map_icd10_to_efo("T88") == set(["complication"])

assert map_icd10_to_efo("C61") == set(["prostate carcinoma"])
assert map_icd10_to_efo("D07") == set(["urogenital neoplasm"])

assert map_icd10_to_efo(None) is None
assert map_icd10_to_efo("NonExistent") is None

In [54]:
_new_df = []

for phecode in test_df["icd10"].unique():
    _new_map_values = map_icd10_to_efo(phecode)
    if _new_map_values is None:
        continue

    for n in _new_map_values:
        _new_df.append({"icd10": phecode, "efo": n})

_new_df = pd.DataFrame(_new_df)

In [55]:
# assign phecode
test_df = pd.merge(test_df, _new_df, on="icd10", how="left")

In [56]:
test_df = test_df.dropna(subset=["icd10", "efo"], how="all")

In [57]:
test_df = test_df.drop_duplicates(subset=["icd10", "efo"])

In [58]:
test_df.shape

(211, 5)

In [59]:
test_df.head()

phecode     phecode_phenotype     phecode_category icd10  \
0     008  Intestinal infection  infectious diseases   A09   
1   008.5   Bacterial enteritis  infectious diseases   A02   
2   008.5   Bacterial enteritis  infectious diseases   A05   
3   008.5   Bacterial enteritis  infectious diseases   A03   
4   008.5   Bacterial enteritis  infectious diseases   A01   

                                   efo  
0  gastroenteritis AND gastroenteritis  
1                                  NaN  
2                                  NaN  
3                                  NaN  
4                                  NaN

## ICD10/EFO to PhenomeXcan traits

In [60]:
assert test_df[pd.isnull(test_df["icd10"])].shape[0] == 0

In [61]:
test_df[pd.isnull(test_df["efo"])]

phecode                             phecode_phenotype  \
1     008.5                           Bacterial enteritis   
2     008.5                           Bacterial enteritis   
3     008.5                           Bacterial enteritis   
4     008.5                           Bacterial enteritis   
8       038                                    Septicemia   
9       038                                    Septicemia   
40      199                Neoplasm of uncertain behavior   
41      202  Cancer of other lymphoid, histiocytic tissue   
54      249                   Secondary diabetes mellitus   
98    285.2                     Anemia of chronic disease   
163   415.2               Chronic pulmonary heart disease   
239     483            Acute bronchitis and bronchiolitis   
240     483            Acute bronchitis and bronchiolitis   
328   707.1                               Decubitus ulcer   

        phecode_category icd10  efo  
1    infectious diseases   A02  NaN  
2    infectious diseases   A05  NaN  
3    infectious diseases   A03  NaN  
4    infectious diseases   A01  NaN  
8    infectious diseases   B00  NaN  
9    infectious diseases   A42  NaN  
40             neoplasms   D20  NaN  
41             neoplasms   C96  NaN  
54   endocrine/metabolic   E12  NaN  
98         hematopoietic   D63  NaN  
163   circulatory system   I27  NaN  
239          respiratory   J20  NaN  
240          respiratory   J05  NaN  
328         dermatologic   L89  NaN

In [62]:
with pd.option_context(
    "display.max_rows", None, "display.max_columns", None, "max_colwidth", None
):
    _tmp = (
        efo_from_icd10.head()
    )  # [efo_from_icd10["current_term_label"] == "osteoarthritis, knee"]
    display(_tmp)

ukb_code              term_label   term_codes mapping_type  \
1       M17  osteoarthritis || knee  EFO:0004616        Broad   
2       R30                 dysuria  EFO:0003901      ? Broad   
7       M05    rheumatoid arthritis  EFO:0000685        Broad   
9       K74      cirrhosis of liver  EFO:0001422        Broad   
10      M15          osteoarthritis  EFO:0002506        Broad   

                                                      ukb_fullcode  \
1      M17-Diagnoses_main_ICD10_M17_Gonarthrosis_arthrosis_of_knee   
2    R30-Diagnoses_main_ICD10_R30_Pain_associated_with_micturition   
7   M05-Diagnoses_main_ICD10_M05_Seropositive_rheumatoid_arthritis   
9     K74-Diagnoses_main_ICD10_K74_Fibrosis_and_cirrhosis_of_liver   
10                      M15-Diagnoses_main_ICD10_M15_Polyarthrosis   

      current_term_label category  
1   osteoarthritis, knee  disease  
2                dysuria      NaN  
7   rheumatoid arthritis  disease  
9     cirrhosis of liver  disease  
10        osteoarthritis  disease

In [63]:
def map_to_phenomexcan_trait(row):
    icd10_list = row["icd10"]
    if pd.isnull(icd10_list):
        icd10_list = set()

    # get EFO from ICD10 hits
    # this has the effect to always find the same set of PhenomeXcan traits
    # regardness of whether only the icd10 or only the efo labels are given)
    icd10_hits = phenomexcan_to_efo[(phenomexcan_to_efo["ukb_code"] == icd10_list)][
        "current_term_label"
    ].tolist()

    efo_list = row["efo"]
    if pd.isnull(efo_list):
        efo_list = set()

    # get ICD10 from EFO hits
    efo_hits = phenomexcan_to_efo[
        (phenomexcan_to_efo["current_term_label"] == efo_list)
    ]["ukb_code"].tolist()

    phenomexcan_maps = phenomexcan_to_efo[
        (phenomexcan_to_efo["ukb_code"].isin([icd10_list] + efo_hits))
        | (phenomexcan_to_efo["current_term_label"].isin([efo_hits] + icd10_hits))
    ]

    if phenomexcan_maps.shape[0] == 0:
        return None

    return set(phenomexcan_maps["ukb_fullcode"].unique())

In [64]:
# some testing
row = {"icd10": "M17", "efo": None}
assert map_to_phenomexcan_trait(row) == set(
    ["M17-Diagnoses_main_ICD10_M17_Gonarthrosis_arthrosis_of_knee"]
)
row = {"icd10": None, "efo": "osteoarthritis, knee"}
assert map_to_phenomexcan_trait(row) == set(
    ["M17-Diagnoses_main_ICD10_M17_Gonarthrosis_arthrosis_of_knee"]
)
row = {"icd10": "M17", "efo": "osteoarthritis, knee"}
assert map_to_phenomexcan_trait(row) == set(
    ["M17-Diagnoses_main_ICD10_M17_Gonarthrosis_arthrosis_of_knee"]
)

row = {"icd10": None, "efo": "coronary artery disease"}
assert map_to_phenomexcan_trait(row) == set(
    [
        "CARDIoGRAM_C4D_CAD_ADDITIVE",
        "I25-Diagnoses_main_ICD10_I25_Chronic_ischaemic_heart_disease",
    ]
)
row = {"icd10": "I25", "efo": None}
assert map_to_phenomexcan_trait(row) == set(
    [
        "CARDIoGRAM_C4D_CAD_ADDITIVE",
        "I25-Diagnoses_main_ICD10_I25_Chronic_ischaemic_heart_disease",
    ]
)
row = {"icd10": "I25", "efo": "coronary artery disease"}
assert map_to_phenomexcan_trait(row) == set(
    [
        "CARDIoGRAM_C4D_CAD_ADDITIVE",
        "I25-Diagnoses_main_ICD10_I25_Chronic_ischaemic_heart_disease",
    ]
)

row = {"icd10": None, "efo": None}
assert map_to_phenomexcan_trait(row) is None

In [65]:
_new_df = []

for idx, row in test_df.iterrows():
    _new_map_values = map_to_phenomexcan_trait(row)
    if _new_map_values is None:
        continue

    for n in _new_map_values:
        _new_df.append({"icd10": row["icd10"], "efo": row["efo"], "phenomexcan": n})

_new_df = pd.DataFrame(_new_df)

In [66]:
assert (
    _new_df[
        _new_df["icd10"].isin(test_df[pd.isnull(test_df["efo"])]["icd10"].tolist())
    ].shape[0]
    == 0
)

In [67]:
# assign phenomexcan trait
test_df = pd.merge(test_df, _new_df, on=["icd10", "efo"], how="left")

In [68]:
_tmp = test_df[test_df["icd10"] == "I25"]
display(_tmp)
assert _tmp.shape[0] == 2

phecode      phecode_phenotype    phecode_category icd10  \
148   411.2  Myocardial infarction  circulatory system   I25   
149   411.2  Myocardial infarction  circulatory system   I25   

                         efo  \
148  coronary artery disease   
149  coronary artery disease   

                                           phenomexcan  
148                        CARDIoGRAM_C4D_CAD_ADDITIVE  
149  I25-Diagnoses_main_ICD10_I25_Chronic_ischaemic...

In [69]:
test_df.shape

(395, 6)

In [70]:
test_df.head()

phecode     phecode_phenotype     phecode_category icd10  \
0     008  Intestinal infection  infectious diseases   A09   
1   008.5   Bacterial enteritis  infectious diseases   A02   
2   008.5   Bacterial enteritis  infectious diseases   A05   
3   008.5   Bacterial enteritis  infectious diseases   A03   
4   008.5   Bacterial enteritis  infectious diseases   A01   

                                   efo  \
0  gastroenteritis AND gastroenteritis   
1                                  NaN   
2                                  NaN   
3                                  NaN   
4                                  NaN   

                                         phenomexcan  
0  A09-Diagnoses_main_ICD10_A09_Diarrhoea_and_gas...  
1                                                NaN  
2                                                NaN  
3                                                NaN  
4                                                NaN

# Save

In [71]:
# FIXME: uncomment when merged with emerge branch
# conf.RESULTS["EMERGE_VALIDATION_DIR"].mkdir(exist_ok=True, parents=True)

In [72]:
# FIXME: hardcoded path
# output_file = Path(
#     conf.RESULTS["EMERGE_VALIDATION_DIR"],
#     "emerge-test_set.pkl",
# ).resolve()
output_file = Path(
    "/home/miltondp/projects/labs/greenelab/phenoplier/base/data/emerge",
    "phecodes_phenomexcan_maps.tsv",
).resolve()

display(output_file)

PosixPath('/home/miltondp/projects/labs/greenelab/phenoplier/bases_data/base_orig/data/emerge/phecodes_phenomexcan_maps.tsv')

In [73]:
test_df.to_csv(output_file, sep="\t", index=False)